language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -82,79 +115,154 @@ mergePackageMap ps = foldM merge Map.empty ps
+             createDirectoryIfMissing True (takeDirectory outputFile)
+             T.writeFileUtf8 outputFile modTxt
+             pure $ ds ++ dependencies
+           Nothing -> pure dependencies
+
+-- Generate the .ts content for a single module.
+genModule :: Map.Map PackageId (Maybe PackageName, Package) ->
+     T.Text -> PackageId -> Module -> Maybe (T.Text, [String])
+genModule pm packageSetName curPkgId mod
+  | null serDefs =
+    Nothing -- If no serializable types, nothing to do.
   | otherwise =
-    let curModName = moduleName mod
-        pkgRootPath
-          | lenModName == 1 = "."
-          | otherwise = T.intercalate "/" (replicate (lenModName - 1) "..")
-          where
-            lenModName = length (unModuleName curModName)
-        tpls = moduleTemplates mod
-        (defSers, refs) = unzip (map (genDataDef curPkgId mod tpls) serDefs)
-        header =
-            ["// Generated from " <> T.intercalate "/" (unModuleName curModName) <> ".daml"
-            ,"/* eslint-disable @typescript-eslint/camelcase */"
-            ,"/* eslint-disable @typescript-eslint/no-use-before-define */"
-            ,"import * as jtv from '@mojotech/json-type-validation';"
-            ,"import * as daml from '@daml/types';"
-            ]
-        imports =
-            ["import * as " <> modNameStr <> " from '" <> pkgRootPath <> "/" <> pkgRefStr <> T.intercalate "/" (unModuleName modName) <> "';"
-            | modRef@(pkgRef, modName) <- Set.toList ((PRSelf, curModName) `Set.delete` Set.unions refs)
-            , let pkgRefStr = case pkgRef of
-                    PRSelf -> ""
-                    PRImport pkgId ->
-                      -- If the imported package has a symbolic name,
-                      -- then we need to use it in the package path.
-                      "../" <> (case Map.lookup pkgId pm of
-                                   Just (Just name, _) -> unPackageName name
-                                   Just (Nothing, _) -> unPackageId pkgId
-                                   Nothing -> error "IMPOSSIBLE : package map malformed"
-                               ) <> "/"
-            , let modNameStr = genModuleRef modRef
-            ]
-        defs = map (\(def, ser) -> def ++ ser) defSers
-    in
-    Just $ T.unlines $ intercalate [""] $ filter (not . null) $ header : imports : defs
+    let (defSers, refs) = unzip (map (genDataDef curPkgId mod tpls) serDefs)
+        imports = [ importDecl pm modRef rootPath
+                  | modRef@(pkgRef, _) <- modRefs refs
+                  , let rootPath = pkgRootPath modName pkgRef ]
+        defs = map biconcat defSers
+        modText = T.unlines $ intercalate [""] $ filter (not . null) $ modHeader : imports : defs
+        depends = [ T.unpack $ packageSetName <> "/" <> pkgRefStr pm pkgRef
+                  | (pkgRef, _) <- modRefs refs, isPkgRefNotSelf pkgRef ]
```

Comment:
```suggestion
                  | (pkgRef, _) <- modRefs refs, pkgRef /= PRSelf]
```

Reply:
i don't know why i didn't think of this :)